pkg/agent/monitortool/monitor.go

```diff
@@ -0,0 +1,449 @@
+	klog.V(4).InfoS("NodeLatencyMonitor updated", "NodeLatencyMonitor", klog.KObj(newNLM))
+
+	if oldNLM.GetGeneration() == newNLM.GetGeneration() {
+		return
+	}
+
+	m.updateLatencyConfig(newNLM)
+}
+
+// updateLatencyConfig updates the latency config based on the NodeLatencyMonitor CRD.
+func (m *NodeLatencyMonitor) updateLatencyConfig(nlm *v1alpha1.NodeLatencyMonitor) {
+	pingInterval := time.Duration(nlm.Spec.PingIntervalSeconds) * time.Second
+
+	m.latencyConfig = &LatencyConfig{
+		Enable:   true,
+		Interval: pingInterval,
+	}
+
+	m.latencyConfigChanged <- struct{}{}
+}
+
+// onNodeLatencyMonitorDelete is the event handler for deleting NodeLatencyMonitor.
+func (m *NodeLatencyMonitor) onNodeLatencyMonitorDelete(obj interface{}) {
+	m.latencyConfig = &LatencyConfig{Enable: false}
+	klog.V(4).InfoS("NodeLatencyMonitor deleted")
+
+	m.latencyConfigChanged <- struct{}{}
+}
+
+// sendPing sends an ICMP message to the target IP address.
+func (m *NodeLatencyMonitor) sendPing(socket net.PacketConn, addr net.IP) error {
+	var requestType icmp.Type
+
+	ip := &net.IPAddr{IP: addr}
+
+	if addr.To4() == nil {
+		requestType = ipv6.ICMPTypeEchoRequest
+	} else {
+		requestType = ipv4.ICMPTypeEcho
+	}
+
+	timeStart := time.Now()
+	seqID := getICMPSeq()
+	body := &icmp.Echo{
+		ID:   icmpEchoID,
+		Seq:  int(seqID),
+		Data: []byte(timeStart.Format(time.RFC3339Nano)),
+	}
+	msg := icmp.Message{
+		Type: requestType,
+		Code: 0,
+		Body: body,
+	}
+	klog.V(4).InfoS("Sending ICMP message", "IP", ip, "SeqID", seqID, "body", body)
+
+	// Serialize the ICMP message
+	msgBytes, err := msg.Marshal(nil)
+	if err != nil {
+		return err
+	}
+
+	// Send the ICMP message
+	_, err = socket.WriteTo(msgBytes, ip)
+	if err != nil {
+		return err
+	}
+
+	// Create or update the latency store
+	mutator := func(entry *NodeIPLatencyEntry) {
+		entry.LastSendTime = timeStart
+	}
+	m.latencyStore.SetNodeIPLatencyEntry(addr.String(), mutator)
+
+	return nil
+}
+
+// recvPing receives an ICMP message from the target IP address.
+func (m *NodeLatencyMonitor) recvPing(socket net.PacketConn, isIPv4 bool) {
+	// We only expect small packets, if we receive a larger packet, we will drop the extra data.
+	readBuffer := make([]byte, 128)
+	for {
+		n, peer, err := socket.ReadFrom(readBuffer)
+		if err != nil {
+			// When the socket is closed in the Run method, this error will be logged, which is not ideal.
+			// In the future, we may try setting a ReadDeadline on the socket before each ReadFrom and using
+			// a channel to signal that the loop should terminate.
+			klog.ErrorS(err, "Failed to read ICMP message")
+			return
+		}
+
+		destIP := peer.String()
+
+		// Parse the ICMP message
+		var msg *icmp.Message
+		if isIPv4 {
+			msg, err = icmp.ParseMessage(ProtocolICMP, readBuffer[:n])
+			if err != nil {
+				klog.ErrorS(err, "Failed to parse ICMP message")
+				continue
+			}
+			if msg.Type != ipv4.ICMPTypeEchoReply {
+				klog.V(4).InfoS("Failed to match ICMPTypeEchoReply")
```

Comment:
I saw a lot of these messages in your recorded demo video. Could you explain which packets are causing this?

Reply:
Hi @Antonin Bas, I think this log is normal.
```bash
Failed to match is mainly the following two cases:
I0518 02:07:40.967464 1 monitor.go:266] "Failed to match ICMPTypeEchoReply" Msg={"Type":3, "Code":3, "Checksum":65316, "Body":{"Data":" RQAAPDVwQABABp4PrBIAAgpgsMjlKAG7sy1P5gAAAACgAvrwZ2sAAAIEBbQEAggKUw6mpAAAAAABAwMH", "Extensions":null}}
```

The equivalent is
```go
ICMPTypeDestinationUnreachable ICMPType = 3 // Destination Unreachable
```

it is what is returned when recvPing is first started and there is no sendPing.

The other case is that the Node in the sendPing contains the Echo sent by the current and other Nodes, i.e. it is received as an Echo and not as a Reply, so there is also a not match, maybe I can filter out the recvPing whose type is ICMPTypeEcho.

```bash
I0518 02:07:46.858090 1 monitor.go:266] "Failed to match ICMPTypeEchoReply" Msg={"Type":8, "Code":0, "Checksum":52950, "Body":{"ID":1, "Seq":3 , "Data": "MjAyNC0wNS0xOFQwMjowNzo0Ni44NTcyMTAyNTda"}}
```

Corresponding to
```go
ICMPTypeEcho ICMPType = 8 // Echo
```

Reply:
Please remove the current Node from the list of IPs to ping. You can use `nodeConfig.Name` as the name of the current Node, and you can modify the 3 Node event handlers (ADD / UPDATE / DELETE) to ignore Node events for which the name matches the current Node name.

Based on your explanation above, let's replace this
```go
			if msg.Type != ipv4.ICMPTypeEchoReply {
				klog.V(4).InfoS("Failed to match ICMPTypeEchoReply", "Msg", msg)
				continue
			}
```
with
```go
			// Ignore ICMP echo messages received from other Nodes (they will be answered by the system)
			if msg.Type == ipv4.ICMPTypeEchoRequest {
				klog.V(7).InfoS("Ignoring ICMPTypeEchoRequest message", "msg", msg)
				continue
			}
			if msg.Type != ipv4.ICMPTypeEchoReply {
				klog.V(5).InfoS("Ignoring all other ICMP messages", "msg", msg)
				continue
			}
```

BTW, I am still not sure why you are receiving a `ICMPTypeDestinationUnreachable` message, and I don't understand your explanation:

> it is what is returned when recvPing is first started and there is no sendPing.

We should only receive this message if there is a network problem. This should have nothing to do with the run order between `recvPing` and sendPing`. This means the other Node cannot be reached. Hopefully you only see this message a few times at the beginning, and not in steady state?

Reply:
Thanks, that looks good to me.

Reply:
The `ICMPTypeDestinationUnreachable` is printed in the first serval seconds, it seems that the coredns pod is in Creating status, when all pods is Running, this `ICMPTypeDestinationUnreachable` is fixed.

Reply:
That should be unrelated. There is no dependency on the coreDNS Pods.